tests/posix/common/src/pthread.c

```diff
@@ -587,12 +587,15 @@ ZTEST(posix_apis, test_pthread_descriptor_leak)
 	pthread_t pthread1;
 	pthread_attr_t attr;
 
+	zassert_ok(pthread_attr_init(&attr));
+	zassert_ok(pthread_attr_setstack(&attr, &stack_e[0][0], STACKS));
+
 	/* If we are leaking descriptors, then this loop will never complete */
 	for (size_t i = 0; i < CONFIG_MAX_PTHREAD_COUNT * 2; ++i) {
-		zassert_ok(pthread_attr_init(&attr));
-		zassert_ok(pthread_attr_setstack(&attr, &stack_e[0][0], STACKS));
 		zassert_ok(pthread_create(&pthread1, &attr, create_thread1, NULL),
 			   "unable to create thread %zu", i);
+		/* Small delay to prevent joining before the thread has been spawned */
```

Comment:
> This looks like a possible source of application bugs if it is actually required here. I would remove this sleep and make sure the implementation was actually capable of doing back-to-back create/join calls.

IIRC, even on Linux there is no guarantee that a thread is spawned before `pthread_create()` returns.

My initial respin of this rev did introduce a semaphore to block the calling thread until the spawned thread had started. However, that would also punish the threads that do not require that bit of synchronization so I removed it.

Reply:
yeah, you definitely don't want to serialize thread startup. However, given that `pthread_create` doesn't return until after `k_thread_create` has been called, I would think it should be valid to call `k_thread_join` any time after `k_thread_create` has returned. If not, then I think we've got some sharp edges in the API that are going to catch developers by surprise.

Reply:
That's indeed the behavior: you can call k_thread_join() at any point after k_thread_create() has returned in any thread, even after it has aborted.   And FWIW: even outside that regime you're safe as long as the k_thread struct is zero-initialized (joining garbage data is, heh, undefined).

Reply:
> yeah, you definitely don't want to serialize thread startup. However, given that `pthread_create` doesn't return until after `k_thread_create` has been called, I would think it should be valid to call `k_thread_join` any time after `k_thread_create` has returned. If not, then I think we've got some sharp edges in the API that are going to catch developers by surprise.

If that should be the expected behaviour, even though it differs from other POSIX systems, then we would likely need to hold a spinlock until after `k_thread_create()` returns.

That might be fine

Reply:
Maybe I'm just misunderstanding something here. I believe, reading the pthreads POSIX spec, that this is valid and well defined code:
```
    pthread_t th;
    pthread_create(&th, NULL, thread_start, NULL);
    pthread_join(th, NULL);
```
Or am I just confused about what we're discussing?

As for holding the spinlock through `k_thread_create()`, I suspect that's a very bad idea as spinlocks will likely block interrupts on the current core, right?

Reply:
From my reading of your current version, I think you'd only need additional synchronization if the underlying `k_thread_create()` and `k_thread_join()` functions aren't safe to invoke back-to-back from the parent thread. You've done a good job of making the pthread implementation a pretty thin wrapper on top of the underlying Zephyr thread implementation, just a bit of state to deal with the extra POSIX semantics.

Reply:
I think it's best to aim for the ideal - so users should feel comfortable that they can busy loop on calling `k_thread_create()` and `k_thread_join()` in quick succession, even if it's not the smartest thing to do.

Unfortunately, the assumptions are not sound. This may require a bit of additional effort on the kernel side.